include/ginkgo/core/base/machine_topology.hpp

```diff
@@ -0,0 +1,420 @@
+
+
+    /**
+     * This struct holds the attributes for an IO/Misc object.
+     *
+     * Mainly used for PCI devices. The identifier important for PCI devices is
+     * the PCI Bus ID, stored here as a string. PCI devices themselves usually
+     * contain Hard disks, network components as well as other objects that are
+     * not important for our case.
+     *
+     * In many cases, hwloc is able to identify the OS devices that belong to a
+     * certain PCI Bus ID and here they are stored in the io children vector. A
+     * list of their names are also additionally stored for easy access and
+     * comparison.
+     *
+     * @note IO children can have names such as ibX for Infiniband cards, cudaX
+     * for NVIDIA cards with CUDA and rsmiX for AMD cards.
+     */
+    struct io_obj_info {
+        /**
+         * The hwloc object.
+         */
+        hwloc_obj_t obj;
+
+        /**
+         * The logical_id assigned by HWLOC (assigned according to proximity).
+         *
+         * @note Use this rather than os_id for all purposes other than binding.
+         * [Reference](https://www.open-mpi.org/projects/hwloc/doc/v2.4.0/a00364.php#faq_indexes)
+         */
+        size_type logical_id;
+
+        /**
+         * The os_id assigned by the OS (assigned arbitrarily by the OS)
+         */
+        size_type os_id;
+
+        /**
+         * The global persistent id assigned to the object by hwloc.
+         */
+        size_type gp_id;
+
+        /**
+         * The closest numa.
+         */
+        int numa;
+
+        /**
+         * The non-io parent object.
+         */
+        hwloc_obj_t non_io_ancestor;
+
+        /**
+         * The ancestor local id.
+         */
+        int ancestor_local_id;
+
+        /**
+         * The ancestor type.
+         */
+        std::string ancestor_type;
+
+        /**
+         * The array of CPU ids closest to the object.
+         */
+        std::vector<int> closest_cpu_ids;
+
+        /**
+         * The PCI Bus ID
+         */
+        std::string pci_bus_id;
+    };
+
+public:
+    /**
+     * Do not allow the MachineTopology object to be copied/moved. There should
+     * be only one global object per execution.
+     */
+    MachineTopology(MachineTopology &) = delete;
+    MachineTopology(MachineTopology &&) = delete;
+    MachineTopology &operator=(MachineTopology &) = delete;
+    MachineTopology &operator=(MachineTopology &&) = delete;
+    ~MachineTopology() = default;
+
+    /**
+     * Bind the object associated with the id to a core.
+     *
+     * @param id  The id of the object to be bound.
+     */
+    void bind_to_cores(const int *ids, const size_type num_ids) const
+    {
+        hwloc_binding_helper(this->cores_, ids, num_ids);
+    }
+
+    /**
+     * Bind the object associated with the id to a Processing unit(PU).
+     *
+     * @param id  The id of the object to be bound.
+     */
+    void bind_to_pus(const int *ids, const size_type num_ids) const
```

Comment:
See previous comments about using `std::vector` and maybe `std::array` instead of pointer-size combination.